# 2.2 Data Preprocessing — Exercises

This notebook works through the exercises from Section 2.2 of *Dive into Deep Learning* (P. 40).

Those questions are :
1. Try loading datasets, e.g., Abalone from the [UCI Machine Learning Repository](https://archive.ics.uci.edu/ml/datasets) and inspect their properties. What fraction of them has missing values? What fraction of the variables is numerical, categorical, or text?
1. Try indexing and selecting data columns by name rather than by column number. The pandas documentation on [indexing](https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html) has further details on how to do this.
1. How large a dataset do you think you could load this way? What might be the limitations? Hint: consider the time to read the data, representation, processing, and memory footprint. Try this out on your laptop. What happens if you try it out on a server? 
1. How would you deal with data that has a very large number of categories? What if the category labels are all unique? Should you include the latter?
1. What alternatives to pandas can you think of? How about [loading NumPy tensors from a file](https://numpy.org/doc/stable/reference/generated/numpy.load.html)? Check out [Pillow](https://python-pillow.org/), the Python Imaging Library.

## Setup

In [2]:
import pandas as pd
import numpy as np

from ucimlrepo import fetch_ucirepo

Some Keybinds (Delete Later):
- Run with Ctrl+Enter.
- Clear Output with Alt+Del.

## 1. Inspecting the Abalone Dataset

We're looking to answer these questions :
- What fraction of the data has missing values?
- What fraction of the variables is numerical?
- What fraction is categorical?
- What fraction is text?


In order to do so, we load the Abalone dataset from the UCI Machine Learning Repository and inspect its properties :

In [ ]:
# Fetch DataSet 
abalone = fetch_ucirepo(id=1) 
  
# Data (as Pandas DataFrames) 
X = abalone.data.features 
y = abalone.data.targets

# Joining the inputs and features in one DataFrame
df = X.join(y)
        

Inspecting the content & elements :

In [ ]:
# metadata 
print("Metadata :")
print(abalone.metadata) 

print('\n\n')

# variable information
print("Variable Infos :") 
print(abalone.variables) 


Metadata :
{'uci_id': 1, 'name': 'Abalone', 'repository_url': 'https://archive.ics.uci.edu/dataset/1/abalone', 'data_url': 'https://archive.ics.uci.edu/static/public/1/data.csv', 'abstract': 'Predict the age of abalone from physical measurements', 'area': 'Biology', 'tasks': ['Classification', 'Regression'], 'characteristics': ['Tabular'], 'num_instances': 4177, 'num_features': 8, 'feature_types': ['Categorical', 'Integer', 'Real'], 'demographics': [], 'target_col': ['Rings'], 'index_col': None, 'has_missing_values': 'no', 'missing_values_symbol': None, 'year_of_dataset_creation': 1994, 'last_updated': 'Mon Aug 28 2023', 'dataset_doi': '10.24432/C55C7W', 'creators': ['Warwick Nash', 'Tracy Sellers', 'Simon Talbot', 'Andrew Cawthorn', 'Wes Ford'], 'intro_paper': None, 'additional_info': {'summary': 'Predicting the age of abalone from physical measurements.  The age of abalone is determined by cutting the shell through the cone, staining it, and counting the number of rings through a mic

In [ ]:
X

In [ ]:
y

In [30]:
df

,Sex,Length,Diameter,Height,Whole_weight,Shucked_weight,Viscera_weight,Shell_weight,Rings
0,M,0.455,0.365,0.095,0.5140,0.2245,0.1010,0.1500,15
1,M,0.350,0.265,0.090,0.2255,0.0995,0.0485,0.0700,7
2,F,0.530,0.420,0.135,0.6770,0.2565,0.1415,0.2100,9
3,M,0.440,0.365,0.125,0.5160,0.2155,0.1140,0.1550,10
4,I,0.330,0.255,0.080,0.2050,0.0895,0.0395,0.0550,7
...,...,...,...,...,...,...,...,...,...
4172,F,0.565,0.450,0.165,0.8870,0.3700,0.2390,0.2490,11
4173,M,0.590,0.440,0.135,0.9660,0.4390,0.2145,0.2605,10
4174,M,0.600,0.475,0.205,1.1760,0.5255,0.2875,0.3080,9
4175,F,0.625,0.485,0.150,1.0945,0.5310,0.2610,0.2960,10


### Dataset shape

In [26]:
df.shape, df.size

((4177, 9), 37593)

The dataset contains **4,177 observations/examples and 9 variables, consisting of 8 input features and 1 target variable** `Rings`.

### Inspect the data

In [33]:
df.head()

,Sex,Length,Diameter,Height,Whole_weight,Shucked_weight,Viscera_weight,Shell_weight,Rings
0,M,0.455,0.365,0.095,0.5140,0.2245,0.1010,0.150,15
1,M,0.350,0.265,0.090,0.2255,0.0995,0.0485,0.070,7
2,F,0.530,0.420,0.135,0.6770,0.2565,0.1415,0.210,9
3,M,0.440,0.365,0.125,0.5160,0.2155,0.1140,0.155,10
4,I,0.330,0.255,0.080,0.2050,0.0895,0.0395,0.055,7


In [6]:
df.tail()

,Sex,Length,Diameter,Height,Whole_weight,Shucked_weight,Viscera_weight,Shell_weight,Rings
4172,F,0.565,0.450,0.165,0.8870,0.3700,0.2390,0.2490,11
4173,M,0.590,0.440,0.135,0.9660,0.4390,0.2145,0.2605,10
4174,M,0.600,0.475,0.205,1.1760,0.5255,0.2875,0.3080,9
4175,F,0.625,0.485,0.150,1.0945,0.5310,0.2610,0.2960,10
4176,M,0.710,0.555,0.195,1.9485,0.9455,0.3765,0.4950,12


In [7]:
df.dtypes

Sex                object
Length            float64
Diameter          float64
Height            float64
Whole_weight      float64
Shucked_weight    float64
Viscera_weight    float64
Shell_weight      float64
Rings               int64
dtype: object

### Missing values

In [24]:
missing_by_column = df.isna().sum()
missing_by_column

Sex               0
Length            0
Diameter          0
Height            0
Whole_weight      0
Shucked_weight    0
Viscera_weight    0
Shell_weight      0
Rings             0
dtype: int64

No missing **values** (Labeled as `NaN`) in this dataset

In [ ]:
# Fraction of all cells that are missing
missing_cell_fraction = df.isna().sum().sum() / df.size

# Fraction of rows containing at least one missing value
missing_row_fraction = df.isna().any(axis=1).mean()

missing_cell_fraction, missing_row_fraction

(0.0, 0.0)

The Abalone dataset contains no missing values, so both fractions are **0%**. This is a bad example to calculate these values

### Variable types

In [28]:
numerical_columns = df.select_dtypes(include="number").columns
object_columns = df.select_dtypes(include="object").columns
categorical_columns = df.select_dtypes(include="category").columns


print("Numerical:", list(numerical_columns))
print("Object/Text:", list(object_columns))
print("Categorical:", list(categorical_columns))


Numerical: ['Length', 'Diameter', 'Height', 'Whole_weight', 'Shucked_weight', 'Viscera_weight', 'Shell_weight', 'Rings']
Object/Text: ['Sex']
Categorical: []


In [29]:
# Quick function to calculate percentage
percentage = lambda x: round(len(x)/len(df.columns), 3)

numerical_fraction = percentage(numerical_columns)
categorical_fraction = percentage(categorical_columns)
object_fraction = percentage(object_columns)

numerical_fraction, categorical_fraction, object_fraction

(0.889, 0.0, 0.111)

Here, `Sex` is represented by Pandas as an `object` column rather than the `category` dtype. But we semantically treat it as a category.

The input variables contain **7 numerical variables and 1 categorical variable**. There are no genuine free-form text variables.

By default, the target `Rings` is numerical and is kept separate in `y` when preparing the supervised learning problem.

## 2. Selecting Columns by Name

Pandas allows columns to be selected by either by labels or by position.
We can use `[]`, `.loc`, `.at`, `.iloc`, `.iat`.

Inspecting the column names:

In [10]:
X.columns

Index(['Sex', 'Length', 'Diameter', 'Height', 'Whole_weight', 'Shucked_weight',
       'Viscera_weight', 'Shell_weight'],
      dtype='object')

Selecting by name is more explicit. `X.iloc[:, 1]` means "the column currently at position 1", while `X["Length"]` means "the column labeled `Length`".

Selecting the second column by its position:

In [11]:
X.iloc[:, 1]

0       0.455
1       0.350
2       0.530
3       0.440
4       0.330
        ...  
4172    0.565
4173    0.590
4174    0.600
4175    0.625
4176    0.710
Name: Length, Length: 4177, dtype: float64

Selecting same column by its label:

In [70]:
X["Length"]

0       0.455
1       0.350
2       0.530
3       0.440
4       0.330
        ...  
4172    0.565
4173    0.590
4174    0.600
4175    0.625
4176    0.710
Name: Length, Length: 4177, dtype: float64

If we needed specific rows (No `X[]` equivalent):

In [69]:
X.loc[1:4, "Length"]

1    0.35
2    0.53
3    0.44
4    0.33
Name: Length, dtype: float64

Multiple columns can be selected by passing a list of labels :

In [59]:
X[["Length", "Diameter", "Height"]]

,Length,Diameter,Height
0,0.455,0.365,0.095
1,0.350,0.265,0.090
2,0.530,0.420,0.135
3,0.440,0.365,0.125
4,0.330,0.255,0.080
...,...,...,...
4172,0.565,0.450,0.165
4173,0.590,0.440,0.135
4174,0.600,0.475,0.205
4175,0.625,0.485,0.150


Same result but with `.loc`

In [ ]:
X.loc[:, ["Length", "Diameter", "Height"]]

,Length,Diameter,Height
0,0.455,0.365,0.095
1,0.350,0.265,0.090
2,0.530,0.420,0.135
3,0.440,0.365,0.125
4,0.330,0.255,0.080
...,...,...,...
4172,0.565,0.450,0.165
4173,0.590,0.440,0.135
4174,0.600,0.475,0.205
4175,0.625,0.485,0.150


The equivalent positional selection is:

In [88]:
X.iloc[:, 1:4]

,Length,Diameter,Height
0,0.455,0.365,0.095
1,0.350,0.265,0.090
2,0.530,0.420,0.135
3,0.440,0.365,0.125
4,0.330,0.255,0.080
...,...,...,...
4172,0.565,0.450,0.165
4173,0.590,0.440,0.135
4174,0.600,0.475,0.205
4175,0.625,0.485,0.150


We pick this table as an example for these below :

In [89]:
X.iloc[:, :4]

,Sex,Length,Diameter,Height
0,M,0.455,0.365,0.095
1,M,0.350,0.265,0.090
2,F,0.530,0.420,0.135
3,M,0.440,0.365,0.125
4,I,0.330,0.255,0.080
...,...,...,...,...
4172,F,0.565,0.450,0.165
4173,M,0.590,0.440,0.135
4174,M,0.600,0.475,0.205
4175,F,0.625,0.485,0.150


Label-based selection is generally easier to maintain because changing the order of the columns does not change which variables are selected.

We can use `X.at[]` or `X.iat[]` if we needed one specific scalar value : 

In [85]:
X.at[2, "Length"], X.iat[2, 1]

(0.53, 0.53)

In [84]:
X.at[15, "Height"], X.iat[15, 3]

(0.13, 0.13)

> **Note:** `X.at[]` uses **labels**, not integer positions. In this dataset, the row labels happen to be `0, 1, 2, ...`, so `X.at[1, "Length"]` can look similar to positional indexing. However, `1` means **the row whose label is `1`**, not necessarily **the second row**.
>
For example, if the row labels were `[10, 20, 30]`:

 ```python
 X.at[20, "Length"]   # row with label 20
 X.iat[1, 0]          # second row, first column
 ```
 Therefore:

 - `.loc` : label-based, can select multiple cells
 - `.iloc` : position-based, can select multiple cells
 - `.at` : label-based, single cell
 - `.iat` : position-based, single cell


## 3. How Large a Dataset Can Be Loaded This Way?

Pandas loads a DataFrame into memory, so the practical limit depends heavily on the available RAM and the representation of the data.

Important factors include:

- **File size:** how much data must be read from disk or downloaded.
- **Memory representation:** the in-memory DataFrame can require considerably more memory than the original CSV.
- **Processing:** operations such as sorting, joining, copying, or converting data can temporarily require additional memory.
- **Available RAM:** the dataset and intermediate objects must fit comfortably in memory.
- **CPU and storage speed:** larger datasets also take longer to read and process.

A useful experiment is to inspect the memory usage of the current dataset.

In [90]:
df.memory_usage(deep=True)

Index                132
Sex               242266
Length             33416
Diameter           33416
Height             33416
Whole_weight       33416
Shucked_weight     33416
Viscera_weight     33416
Shell_weight       33416
Rings              33416
dtype: int64

Footprint in MiB:

In [23]:
df.memory_usage(deep=True).sum() / 1024**2, "MiB"

(0.4861125946044922, 'MiB')

There is no single maximum dataset size. A machine with 8 GB of RAM and a machine with 128 GB of RAM have very different practical limits. Furthermore, a DataFrame that occupies 4 GB should not be expected to work comfortably on a machine with only 4 GB of available RAM because Pandas may need additional memory for intermediate operations.

On a server, the same principle applies, but a server may have substantially more RAM and CPU resources. For datasets that exceed the resources of a single machine, distributed or out-of-core tools such as Dask or Spark can be considered.

## 4. Variables With a Very Large Number of Categories

A categorical variable with 100,000 categories could create 100,000 additional columns, It can make the representation very large and can make model training more expensive.

Inspecting the number of categories in the Abalone `Sex` variable :

In [24]:
X["Sex"].nunique()

3

In [25]:
X["Sex"].value_counts()

Sex
M    1528
I    1342
F    1307
Name: count, dtype: int64

For a variable with many categories, possible approaches include:

- grouping rare categories into an `Other` category;
- using frequency or count encoding;
- using target encoding when appropriate and carefully avoiding data leakage;
- using learned embeddings, especially for neural networks;
- using models that can handle categorical variables directly;
- dropping the variable when it provides little useful information.

### What if every category is unique?

If every row has a unique category, one-hot encoding essentially creates one column per observation. This usually does not provide useful generalizable information.

For example, an identifier such as:

```text
user_000001
user_000002
user_000003
...
```

should normally not be treated as a useful categorical feature merely because it contains strings. It is often an identifier rather than a meaningful predictive variable.

Therefore, a column whose values are all unique should be investigated before being included in a model.

## 5. Alternatives to Pandas

Pandas is useful for tabular data, but it is not the only option.

Some alternatives and related tools include:

| Tool | Typical use |
|---|---|
| NumPy | Numerical arrays and mathematical operations |
| Polars | Fast tabular data processing |
| Dask | Larger-than-memory and parallel DataFrame workloads |
| Apache Spark | Distributed processing of very large datasets |
| PyArrow | Columnar data and efficient data interchange |
| Pillow | Loading and processing images |

For numerical arrays, NumPy can load arrays directly from files:

In [19]:
# Example syntax for a NumPy .npy file:
# array = np.load("array.npy")

For images, Pillow provides image-loading functionality:

In [20]:
# Example syntax:
# from PIL import Image
# image = Image.open("image.jpg")

The appropriate tool depends on the structure and scale of the data. Pandas is particularly convenient for ordinary tabular datasets, while NumPy is more focused on numerical arrays and tools such as Spark are designed for distributed processing.